libs/pageserver_api/src/keyspace.rs

```diff
@@ -27,6 +27,55 @@ impl std::fmt::Display for KeySpace {
     }
 }
 
+pub struct KeySpaceIter<'a> {
+    ranges: &'a [Range<Key>],
+    current_range: usize,
+    current_key: Option<Key>,
+}
+
+impl<'a> Iterator for KeySpaceIter<'a> {
+    // (current_key, range_end)
+    type Item = (Key, Key);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // if we've gone through all ranges, stop iteration
+        if self.current_range >= self.ranges.len() {
+            return None;
+        }
+
+        let range = &self.ranges[self.current_range];
+
+        // if current_key is None, initialize it with range.start
+        if self.current_key.is_none() {
+            self.current_key = Some(range.start);
+        }
+
+        if let Some(current_key) = self.current_key {
```

Comment:
`self.current_key` is always `Some` here, so the nesting is a bit confusing.
I'd do something like:
```
let current_key = match self.current_key {
  None => range.start,
  Some(k) => k,
};

...
```